### PR TITLE
Build MSI packages

### PR DESCRIPTION
**Pull request by smarsching**
## Description
This PR adds code to `CMakeLists.txt` for building a Windows Installer package (MSI) using the WiX toolkit.

## Motivation and context
MSI packages are preferable when automatically deploying software on (many) Windows PCs. Many software management solutions support MSI and there are standardized ways for making local changes to packages (e.g. adding or removing start menu or desktop shortcuts). This is why it is useful to have MSI packages in addition to the NSIS-based installer.

## How has this been tested?
I have tested that the build process still runs correctly with the changes made to `CMakeLists.txt`. I have also tested that the NSIS-based installer is not affected by these changes and still works correctly.

In addition to that, I have verified that installation and deinstallation of the built MSI package. In order to be sure that there are no problems in the future, I also checked that a newer MSI package (having a greater version number) can be installed when an older version is already present.

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ❌ All new and existing tests passed. **[REQUIRED]**
   `testcsvparser` fails, but it also fails without my changes being present.
- ❌ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
  As this PR does not make any changes to the actual code, it cannot introduce new problems.

## Notes:
In order for the MSI package to be built, the WiX toolset is needed. This means that the WiX toolset has to be installed and either its `bin`directory has to be in the `PATH` or the `WIX` environment variable has to point to its root directory.

### CODE REVIEW

**Comment by hifi** [comment]
Wouldn't it make sense to ship only MSI installers? Is there a caveat?

**Comment by smarsching** [comment]
While I like the idea of replacing the NSIS-based installer with an MSI-based one, there are three problems which would need to be solved first:

1. **Feature parity:**
The NSIS-based installer allows the user to choose the name of the group created in the start menu. The MSI-based installer does not (in fact it does not create a group, but simply adds the (single) shortcut directly to the root of the start menu. I do not think that this is a major issue and one could probably live without this feature. Even if we decided that this feature is needed, it should be possible to add it to the MSI-based installer.

2. **Localization:**
At the moment, the MSI-based installer is not localized. This means that the installer’s GUI will always be in English, regardless of the language of the system it runs on. This only affects the installer, localization of the application is not an issue. I have never created a localized MSI-based installer. It should be possible, but it makes the build process much more complicated, because the only way to do this seems to be creating a separate MSI package for each locale and then merging the packages in a later step (see https://www.hass.de/content/how-create-msi-packages-multilingual-user-interface-mui).
As long as the MSI-based installer is simply an extra option, I do not think that localization is an issue: It will primarily be used by system administrators which want to automate the installation and they will typically run it in non-interactive mode anyway. However, if it is the only installer that we offer, we probably want it to be localized.

3. **Upgrading older versions that have been installed using the NSIS installer:**
At the moment, the MSI-based installer will only detect if an older version of KeePassXC has been installed through MSI. It cannot detect a version that has been installed using the NSIS-based installer. This is a severe problem because installing the MSI over a version that has been installed using NSIS will cause inconsistencies (both versions will appear in Add/Remove Programs, but only the one that has been installed second will work).
In theory, it is possible to avoid this problem by using a custom action for detecting an older version that has been installed through NSIS. However, writing such a custom action (which has to be supplied in the form of a DLL) requires significant effort. There are quite a few [“Standard Custom Actions”](http://wixtoolset.org/documentation/manual/v3/customactions/) already distributed as part of WiX, but none of them seems particularly well-suited for our purpose.
It might be possible to somehow do it with the [“Quiet Execution Custom Action”](http://wixtoolset.org/documentation/manual/v3/customactions/qtexec.html) that allows for running an arbitrary command-line, but someone would have to figure out the details.
As long as the MSI-based installer is only used by system administrators, I do not think that it is a major issue that NSI-based installations are not detected because in an automated environment, the software distribution system will typically take care of first uninstalling the old version.

In short: If we want to completely replace the NSIS-based installer we have to resolve the third point and we should probably also resolve the second point. Both are non-trivial problems and most likely require some time (and expertise) to be solved. When we offer the MSI packages as an additional option for system administrators (and clearly label them as such) both issues become far less critical and thus system administrators can benefit from having MSI packages without us having to invest a lot of time. Even if we decide at a later point that we want to replace the NSIS-based installer completely, we can always add the missing features to the MSI-based installer at this later point in time.

**Comment by smarsching** [comment]
@phoerious Sounds like a good idea to me. When building repeatedly for testing purposes, building the complete installer might be a waste of time.

What do you think about making the NSIS installer optional, too? When I did my tests, building the NSIS installer actually took longer than building the MSI, and like the MSI, the right tools need to be available in the `PATH` in order for the build to succeed.

On the other hand, one could argue that it is always possible to limit the type of installers built by specifying the `-G` parameter when calling `cpack`. So introducing configuration flags to control which installers are built feels kind of redundant.

**Comment by droidmonkey** [comment]
I prefer just to use the already existing cpack command line parameters

**Comment by phoerious** [comment]
Is it possible then still to disable MSI by default?

**Comment by droidmonkey** [comment]
You could just not include MSI in ```set(CPACK_GENERATOR "ZIP;NSIS")``` and build MSI using cpack -G MSI

**Comment by phoerious** [comment]
If it works like that, I would prefer it.

**Comment by smarsching** [comment]
Okay, so do we agree, that we only build the ZIP and NSIS installer by default and only build the MSI when explicitly requests via `cpack -G MSI`?

If so, I will push the corresponding change. I only hope that building the MSI packages as part of the normal release process is not forgotten in that case. 🤔

**Comment by phoerious** [comment]
Add it as a side note to the build documentation in the wiki.

**Comment by droidmonkey** [comment]
The release tool is used for release builds, that is where cpack is called with the desired parameters.

**Comment by smarsching** [comment]
I changed the `CPACK_GENERATOR` variable back to `ZIP;NSIS`. This means that the MSI packages will not be built by default, but only when explicitly selected.

However, I did not find a good way of enabling the `WIX` generator when running the release tool: The release tool does not call `cpack` directly, but instead calls `make package`. This means that there is no way of specifying the "-G" parameter when calling CPack. Setting the `CPACK_GENERATOR` variable when calling `cmake` does not help either. The value from `CMakeLists.txt` seems to override the value specified on the command line. This means that `CPackConfig.cmake` will contain the wrong (default) value and the MSI package is not going to be built.

As far as I can tell the only way to fix this in the `release-tool` script is not calling `make package`, but `make preinstall`, followed by `cpack -G "ZIP;NSIS;WIX" --config ./CPackConfig.cmake`. In theory, this should have the same effect as calling `make package`, but maybe someone has an idea for a nicer way how to do this.

**Comment by droidmonkey** [comment]
Changing the command makes sense to me

**Comment by smarsching** [comment]
I pushed a change for the `release-tool` so that now we first call `make preinstall` followed by `cpack -G "NSIS;WIX;ZIP"` instead of calling `make package`. I also added the `.msi` extension to the `mv` command.

I tested the changes: `release-tool` builds the MSI and moves it into the `release` directory.

**Comment by phoerious** [comment]
Looks good. Please update the branch, so I can merge. Thanks!

**Comment by smarsching** [comment]
@phoerious I updated the branch, you should be able to merge it now.

**Comment by phoerious** [comment]
Thanks!
